server/src/main/java/org/apache/druid/server/QueryResource.java

```diff
@@ -171,161 +167,112 @@ public Response cancelQuery(@PathParam("id") String queryId, @Context final Http
+      // Returning a Response object from this point forward requires that object to know the status code, which we
+      // don't actually know until we are in the accumulator, but if we try to return a Response object from the
+      // accumulator, we cannot properly stream results back, because the accumulator won't release control of the
+      // Response until it has consumed the underlying Sequence.
+      final AsyncContext asyncContext = req.startAsync();
 
       try {
-        final ObjectWriter jsonWriter = queryLifecycle.newOutputWriter(ioReaderWriter);
-
-        Response.ResponseBuilder responseBuilder = Response
-            .ok(
-                new StreamingOutput()
-                {
-                  @Override
-                  public void write(OutputStream outputStream) throws WebApplicationException
-                  {
-                    Exception e = null;
-
-                    CountingOutputStream os = new CountingOutputStream(outputStream);
-                    try {
-                      // json serializer will always close the yielder
-                      jsonWriter.writeValue(os, yielder);
-
-                      os.flush(); // Some types of OutputStream suppress flush errors in the .close() method.
-                      os.close();
-                    }
-                    catch (Exception ex) {
-                      e = ex;
-                      log.noStackTrace().error(ex, "Unable to send query response.");
-                      throw new RuntimeException(ex);
-                    }
-                    finally {
-                      Thread.currentThread().setName(currThreadName);
-
-                      queryLifecycle.emitLogsAndMetrics(e, req.getRemoteAddr(), os.getCount());
-
-                      if (e == null) {
-                        successfulQueryCount.incrementAndGet();
-                      } else {
-                        failedQueryCount.incrementAndGet();
-                      }
-                    }
-                  }
-                },
-                ioReaderWriter.getResponseWriter().getResponseType()
-            )
-            .header(QUERY_ID_RESPONSE_HEADER, queryId);
-
-        attachResponseContextToHttpResponse(queryId, responseContext, responseBuilder, jsonMapper,
-                                            responseContextConfig, selfNode
-        );
-
-        return responseBuilder.build();
+        new QueryResourceResultPusher(req, queryLifecycle, io, (HttpServletResponse) asyncContext.getResponse())
+            .push();
       }
-      catch (QueryException e) {
-        // make sure to close yielder if anything happened before starting to serialize the response.
-        yielder.close();
+      finally {
+        asyncContext.complete();
+      }
+    }
+    catch (Exception e) {
+      if (e instanceof ForbiddenException && !req.isAsyncStarted()) {
```

Comment:
that's throwing the exception, not catching it.  Also, that's throwing it from inside of the try, this catch is of generic Exception, which will catch the exception thrown from inside of its own try.  This is basically ensuring that we pass it through.  It's ugly, I agree, but ultimately the mismatch is that we expect all exceptions to be handled and converted into responses here, except for this one exception that we expect to make it out of the method.  We should really refactor it so that we serialize and handle the ForbiddenException in this code as well, which will make it all consistent.  That is, instead of throwing the ForbbidenException, the code above could just create the proper Response object and return it.